src/activity_actor.cpp

```diff
@@ -1292,6 +1293,234 @@ std::unique_ptr<activity_actor> try_sleep_activity_actor::deserialize( JsonIn &j
+    }
+    // free training requires all limbe intact, but specialized workout machines
+    // train upper or lower parts of body only and may permit workout with
+    // broken limbs as long as they are not involved by the machine
+    bool hand_equipment = g->m.has_flag_furn( "WORKOUT_ARMS", location );
+    bool leg_equipment = g->m.has_flag_furn( "WORKOUT_LEGS", location );
+    if( hand_equipment && ( ( who.is_limb_broken( hp_arm_l ) ) ||
+                            who.is_limb_broken( hp_arm_r ) ) ) {
+        who.add_msg_if_player( _( "You cannot train here with a broken arm." ) );
+        act_id = activity_id::NULL_ID();
+        act.set_to_null();
+        return;
+    }
+    if( leg_equipment && ( ( who.is_limb_broken( hp_leg_l ) ) ||
+                           who.is_limb_broken( hp_leg_r ) ) ) {
+        who.add_msg_if_player( _( "You cannot train here with a broken leg." ) );
+        act_id = activity_id::NULL_ID();
+        act.set_to_null();
+        return;
+    }
+    if( !hand_equipment && !leg_equipment &&
+        ( who.is_limb_broken( hp_arm_l ) ||
+          who.is_limb_broken( hp_arm_r ) ||
+          who.is_limb_broken( hp_leg_l ) ||
+          who.is_limb_broken( hp_leg_r ) ) ) {
+        who.add_msg_if_player( _( "You cannot train freely with a broken limb." ) );
+        act_id = activity_id::NULL_ID();
+        act.set_to_null();
+        return;
+    }
+    uilist workout_query;
+    workout_query.desc_enabled = true;
+    workout_query.text =
+        _( "Physical effort determines workout efficiency, but also rate of exhaustion." );
+    workout_query.title = _( "Choose training intensity:" );
+    workout_query.addentry_desc( 1, true, 'l', _( "Light" ),
+                                 _( "Light excercise comparable in intensity to walking, but more focused and methodical." ) );
+    workout_query.addentry_desc( 2, true, 'm', _( "Moderate" ),
+                                 _( "Moderate excercise without excessive exertion, but with enough effort to break a sweat." ) );
+    workout_query.addentry_desc( 3, true, 'a', _( "Active" ),
+                                 _( "Active excercise with full involvement.  Strenuous, but in a controlled manner." ) );
+    workout_query.addentry_desc( 4, true, 'h', _( "High" ),
+                                 _( "High intensity excercise with maximum effort and full power.  Exhausting in the long run." ) );
+    workout_query.query();
+    switch( workout_query.ret ) {
+        case UILIST_CANCEL:
+            act.set_to_null();
+            act_id = activity_id::NULL_ID();
+            return;
+        case 4:
+            act_id = activity_id( "ACT_WORKOUT_HARD" );
+            intensity_modifier = 4;
+            break;
+        case 3:
+            act_id = activity_id( "ACT_WORKOUT_ACTIVE" );
+            intensity_modifier = 3;
+            break;
+        case 2:
+            act_id = activity_id( "ACT_WORKOUT_MODERATE" );
+            intensity_modifier = 2;
```

Comment:
I just wonder why you need different activity ids for the different intensity levels. You already have `intensity_modifier` to record the intensity. Why don't you use that one and use a generic workout activity id all the time.

This would also make the comparison in `player_activity::do_turn` a bit simpler.

Reply:
AFAIK because activity level is JSONized on activity type level (`"activity_level": "EXTRA_EXERCISE"`) so I went with `p.increase_activity_level( actor->get_type()->get_activity_level() );`  in `player_activity::do_turn`, dynamic `actor->get_type()` and `sychronize_type_with_actor();` to make `player_activity` on the same level as current state of the mutable `actor`.

I don't know if this "the way", but this solution works.

Reply:
Well, you could just serialize `intensity_modifier`, like the rest of the properties of `workout_activity_actor`. Maybe that helps. It is currently not serialized. The value of `rest_mode` is also not serialized.